File: dsa_leetcode_beginner/0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists or len(lists) == 0:
            return None
        if len(lists) == 1:
            return lists[0]

        # Divide the list of linked lists into two halves and merge each half recursively
        mid = len(lists) // 2
        left = self.mergeKLists(lists[:mid])
        right = self.mergeKLists(lists[mid:])

        # Merge the two halves
        return self.mergeTwoLists(left, right)

    
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode()
        node = dummy

        # Merge both lists while there are still nodes in l1 and l2
        while l1 and l2:
            if l1.val < l2.val:
                node.next = l1
                l1 = l1.next
            else:
                node.next = l2
                l2 = l2.next
            node = node.next

        # If one list is exhausted, append the other list
        node.next = l1 or l2
        
        return dummy.next
    

#Question: https://leetcode.com/problems/merge-k-sorted-lists
#Blog: https://blog.unwiredlearning.com/merge-k-sorted-lists

File: dsa_leetcode_beginner/0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
#include <vector>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) {
            return nullptr;
        }
        if (lists.size() == 1) {
            return lists[0];
        }

        int mid = lists.size() / 2;
        vector<ListNode*> leftLists(lists.begin(), lists.begin() + mid);
        vector<ListNode*> rightLists(lists.begin() + mid, lists.end());

        ListNode* left = mergeKLists(leftLists);
        ListNode* right = mergeKLists(rightLists);

        return mergeTwoLists(left, right);
    }

private:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode dummy;
        ListNode* node = &dummy;

        while (l1 && l2) {
            if (l1->val < l2->val) {
                node->next = l1;
                l1 = l1->next;
            } else {
                node->next = l2;
                l2 = l2->next;
            }
            node = node->next;
        }

        node->next = l1 ? l1 : l2;
        return dummy.next;
    }
};
